### PR TITLE
Add ability to run collection via UID reference

### PR DESCRIPTION
**Pull request by codenirvana**
**Behavior:**
- Pass API key using `--postman-api-key` flag.
- External resources can be loaded using UID.
  - Check for a valid UID first.
  - Check if no file exists with the same UID name.
  - Finally, check if `postmanApiKey` is present.

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -81,21 +83,36 @@ util = {
...
+            endpoint;
+
+        // build API URL if `location` is a valid UID and api key is provided.
+        // Fetch from file in case a file with valid UID name is present.
+        if (postmanApiKey && UID_REGEX.test(location) && !fs.existsSync(location)) {
+            endpoint = (type === 'collection' ? 'collections' : 'environments') + '/' + location;
+            location = `https://${PRO_API_HOST}/${endpoint}`;
+            headers[API_KEY_HEADER] = postmanApiKey;
+        }
+
```
Comment (kunagpal) [inline lib/util.js]: Extra blank line.

File: lib/util.js [hunk 2]
```diff
@@ -81,21 +83,36 @@ util = {
...
+    fetchJson: function (type, location, options, callback) {
         !callback && _.isFunction(options) && (callback = options, options = {});
 
+        var postmanApiKey = _.get(options, 'postmanApiKey'),
+            headers = { 'User-Agent': USER_AGENT_VALUE },
+            endpoint;
+
+        // build API URL if `location` is a valid UID and api key is provided.
+        // Fetch from file in case a file with valid UID name is present.
+        if (postmanApiKey && UID_REGEX.test(location) && !fs.existsSync(location)) {
```
Comment (kunagpal) [inline lib/util.js]: Make `!fs.existsSync` the first check.

File: test/library/postman-api-key.test.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+            let requestArg = request.get.firstCall.args[0];
+
+            expect(requestArg).to.be.an('object').with.keys(['url', 'json', 'headers']);
+
+            expect(requestArg.url).to.be.a('string')
+                .that.is.equal('https://api.getpostman.com/collections/1234-588025f9-2497-46f7-b849-47f58b865807');
+
+            expect(requestArg.headers).to.be.an('object')
+                .that.has.property('X-Api-Key').to.equal('12345678');
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: The expected value can be passed as the second argument along with `X-Api-Key`.

File: test/library/postman-api-key.test.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+            expect(summary).to.be.an('object')
+                .that.has.property('collection').to.be.an('object')
+                .and.that.include({ id: 'C1', name: 'Collection' });
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: `.and.include`

File: test/library/postman-api-key.test.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        });
+    });
+
+    it('should fetch environment via UID', function (done) {
+        newman.run({
+            collection: 'test/fixtures/run/single-get-request.json',
+            environment: '1234-931c1484-fd1e-4ceb-81d0-2aa102ca8b5f',
+            postmanApiKey: '12345678'
+        }, function (err, summary) {
+            expect(err).to.be.null;
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: `not.be.ok`

File: test/library/postman-api-key.test.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            expect(request.get.calledOnce).to.be.true;
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: sinon.assert.calledOnce(request.get);

File: test/library/postman-api-key.test.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+            let requestArg = request.get.firstCall.args[0];
+
+            expect(requestArg).to.be.an('object').with.keys(['url', 'json', 'headers']);
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: Can be replaced with an includes assertion.

File: test/library/postman-api-key.test.js [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        });
+    });
+
+    it('should fetch globals via UID', function (done) {
+        newman.run({
+            collection: 'test/fixtures/run/single-get-request.json',
+            globals: '1234-6863abf8-6630-4eec-b9cc-2a58f5efe589',
+            postmanApiKey: '12345678'
+        }, function (err, summary) {
+            expect(err).to.be.null;
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: See comments above.

File: test/library/postman-api-key.test.js [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+            done();
+        });
+    });
+
+    it('should fetch all resources via UID', function (done) {
+        newman.run({
+            collection: '1234-588025f9-2497-46f7-b849-47f58b865807',
+            environment: '1234-931c1484-fd1e-4ceb-81d0-2aa102ca8b5f',
+            globals: '1234-6863abf8-6630-4eec-b9cc-2a58f5efe589',
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: The Postman API doesn't have a globals endpoint, this test creates confusion and should be removed.

File: test/library/postman-api-key.test.js [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+            done();
+        });
+    });
+
+    it('should throw error without postmanApiKey', function (done) {
+        newman.run({
+            collection: '1234-588025f9-2497-46f7-b849-47f58b865807'
+        }, function (err) {
+            expect(err).to.be.ok.that.match(/no such file or directory/);
+            expect(request.get.called).to.be.false;
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: Switch to sinon assertions.

File: test/library/postman-api-key.test.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            postmanApiKey: '12345678'
+        }, function (err, summary) {
+            expect(err).to.be.null;
+            expect(request.get.calledThrice).to.be.true;
+
+            expect(summary).to.be.an('object').with.keys(['collection', 'environment', 'globals', 'run']);
+
+            expect(summary.collection).to.include({ id: 'C1', name: 'Collection' });
+            expect(summary.environment).to.include({ id: 'V1', name: 'Variable' });
+            expect(summary.globals).to.include({ id: 'V1', name: 'Variable' });
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: Assert on run as well.

File: test/library/postman-api-key.test.js (hunk 7 above)
Comment (codenirvana) [reply test/library/postman-api-key.test.js]: `includes` doesn't work with objects.

File: lib/util.js [hunk 12]
```diff
@@ -20,7 +20,9 @@ var fs = require('fs'),
     FILESIZE_OPTIONS = { spacer: '' },
 
     PRO_API_HOST = 'api.getpostman.com',
```
Comment (kunagpal) [inline lib/util.js]: `POSTMAN_API_HOST`

File: test/library/postman-api-key.test.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            postmanApiKey: '12345678'
+        }, function (err, summary) {
+            expect(err).to.be.null;
+            sinon.assert.calledOnce(request.get);
+
+            let requestArg = request.get.firstCall.args[0];
+
+            expect(requestArg).to.be.an('object').with.keys(['url', 'json', 'headers']);
+
+            expect(requestArg.url).to.be.a('string')
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: Directly comparing the value will suffice.

File: test/library/postman-api-key.test.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            environment: '1234-931c1484-fd1e-4ceb-81d0-2aa102ca8b5f',
+            postmanApiKey: '12345678'
+        }, function (err, summary) {
+            expect(err).to.be.null;
+            sinon.assert.calledTwice(request.get);
+
+            expect(summary).to.be.an('object').with.keys(['collection', 'environment', 'globals', 'run']);
+
+            expect(summary.collection).to.include({ id: 'C1', name: 'Collection' });
+            expect(summary.environment).to.include({ id: 'E1', name: 'Environment' });
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: Need assertions on the run object as well.

File: lib/util.js [hunk 15]
```diff
@@ -81,21 +83,35 @@ util = {
...
 
+        var postmanApiKey = _.get(options, 'postmanApiKey'),
+            headers = { 'User-Agent': USER_AGENT_VALUE },
+            endpoint;
+
+        // build API URL if `location` is a valid UID and api key is provided.
+        // Fetch from file in case a file with valid UID name is present.
+        if (!fs.existsSync(location) && postmanApiKey && UID_REGEX.test(location)) {
+            endpoint = (type === 'collection' ? 'collections' : 'environments') + '/' + location;
+            location = `https://${POSTMAN_API_HOST}/${endpoint}`;
```
Comment (kunagpal) [inline lib/util.js]: Shift the `https` to the `const`.

File: lib/util.js [hunk 16]
```diff
@@ -81,21 +83,35 @@ util = {
...
         !callback && _.isFunction(options) && (callback = options, options = {});
 
+        var postmanApiKey = _.get(options, 'postmanApiKey'),
+            headers = { 'User-Agent': USER_AGENT_VALUE },
+            endpoint;
+
+        // build API URL if `location` is a valid UID and api key is provided.
+        // Fetch from file in case a file with valid UID name is present.
+        if (!fs.existsSync(location) && postmanApiKey && UID_REGEX.test(location)) {
+            endpoint = (type === 'collection' ? 'collections' : 'environments') + '/' + location;
```
Comment (kunagpal) [inline lib/util.js]: This needs to be made more generic.

File: lib/util.js [hunk 17]
```diff
@@ -19,9 +19,24 @@ var fs = require('fs'),
      */
     FILESIZE_OPTIONS = { spacer: '' },
 
+    POSTMAN_API_PROTOCOL = 'https',
+
     POSTMAN_API_HOST = 'api.getpostman.com',
```
Comment (kunagpal) [inline lib/util.js]: Rename this to `POSTMAN_API_URL` and change the value to `https://api.getpostman.com`

File: lib/util.js [hunk 18]
```diff
@@ -19,8 +19,25 @@ var fs = require('fs'),
...
+    POSTMAN_API_PATH = {
+        collection: 'collections',
+        environment: 'environments'
+    },
+
+    API_KEY_HEADER = 'X-Api-Key',
+
+    USER_AGENT_VALUE = 'Newman/' + version,
+
+    UID_REGEX = /^[0-9A-Z]+-[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}$/i;
```
Comment (kunagpal) [inline lib/util.js]: Check if the API handles IDs in a case-insensitive manner.

File: test/library/postman-api-key.test.js [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+            collection: '1234-588025f9-2497-46f7-b849-47f58b865807'
+        }, function (err) {
+            expect(err).to.be.ok.that.match(/no such file or directory/);
+            sinon.assert.notCalled(request.get);
+
+            done();
+        });
+    });
+
+    it('should not pass API Key header with URL ', function (done) {
```
Comment (kunagpal) [inline test/library/postman-api-key.test.js]: Test title is incomplete. Also, add a test (if not already present) for cases where the API Key should be sent with regular Postman API URls.

File: lib/util.js (hunk 18 above)
Comment (codenirvana) [reply lib/util.js]: Yes, it has `ignoreCase` flag.

**Pull request merged**